src/service-worker.ts

```diff
@@ -48,4 +58,33 @@ registerRoute(({ request, url }: { request: Request; url: URL }) => {
   }
 
   return true
-}, createHandlerBoundToURL(process.env.PUBLIC_URL + '/index.html'))
+}
+
+const APP_SHELL_HTML = process.env.PUBLIC_URL + '/index.html'
+let cachedAppShell: Response | undefined
+
+async function handleAppShell({ request, event }: RouteHandlerCallbackOptions) {
+  // NB: If the cached response is used, the preloadResponse may be cancelled before it is settled.
+  // This will log an error to the console, but it can be ignored - it *should* be cancelled.
+  const { preloadResponse } = event as unknown as { preloadResponse: Promise<Response | undefined> }
+  const controller = new AbortController()
+  const response = (await preloadResponse) || (await fetch(APP_SHELL_HTML, { signal: controller.signal }))
+
+  // Only serve from the cache if the cache is still fresh (checked using etag comparison). The etag
+  // header can be queried before the response has streamed, so this still reduces network time.
+  // (In manual testing, etag lookup took only 2-20% of the total response download time.)
+  const etag = response.headers.get('etag')
```

Comment:
So this doesn't wait for the whole response to be fetched? It just waits for the headers to come back? Or how does this come back faster than the actual page?

Reply:
This JavaScript API is poorly documented in terms of its performance characteristics. 

HTTP responses are text, starting with the headers, then the body, then trailing headers. Response bodies are streamed (by JS APIs), and are only available as promises off of the settled Response promise.

I'm making an assumption, based on this, that the headers are available before the body (because it _is_ available first, to the underlying program, due to the way the response is received over tcp/http). I verified this assumption by timing how long it took to read the etag header vs how long it took to read the full response, and saw that the etag header was available in 20% of the time.